nixos/modules/services/security/vault.nix

```diff
@@ -0,0 +1,208 @@
+{ config, lib, pkgs, utils, ... }:
+
+with lib;
+let
+
+  inherit (pkgs) vault;
+
+  cfg = config.services.vault;
+ 
+  configFile = pkgs.writeText "vault.hcl" ''
+    listener "tcp" {
+      address = "${cfg.listener.address}"
+   
+      ${optionalString (cfg.listener.cluster_address != null)''
+        cluster_address = "${cfg.listener.cluster_address}"
+      ''}
+      
+      ${optionalString (cfg.listener.tls_cert_file != null)''
+        tls_cert_file = "${cfg.listener.tls_cert_file}"
+      ''}
+      
+      ${optionalString (cfg.listener.tls_key_file != null)''
+        tls_key_file = "${cfg.listener.tls_key_file}"
+      ''}
+      
+      ${if cfg.listener.tls_disable then "tls_disable = \"1\"" else "" } 
+
+      tls_min_version = "${cfg.listener.tls_min_version}" 
+
+
+      ${optionalString (cfg.listener.tls_cipher_suites != null)''
+        tls_cipher_suites = \"${cfg.listener.tls_cipher_suites}\"
+      ''}
+
+      ${if cfg.listener.tls_prefer_server_cipher_suites  then "tls_prefer_server_cipher_suites = \"true\"" else "tls_prefer_server_cipher_suites = \"false\"" } 
+
+      ${if cfg.listener.tls_require_and_verify_client_cert then "tls_require_and_verify_client_cert = \"true\"" else "tls_require_and_verify_client_cert = \"false\"" } 
```

Comment:
You could use:

    tls_require_and_verify_client_cert = "${boolToString cfg.listener.tls_require_and_verify_client_cert}";

or even abstract it some more and use it on more places:

    let boolOption = name: opt: "${name} = \"${boolToString opt}\"";
    in ''
      ${boolOption "tls_prefer_server_cipher_suites" cfg.listener.tls_require_and_verify_client_cert}
    '';

So you don't have to repeat yourself. Same with the string options. :)